T39195 Refactor kcidb bridge service

JenySadadia opened the pull request:
Refactor KCIDB bridge service for the below improvements:

- use `Service` as a base class
- code refactoring by defining methods
- define configurations options for KCIDB project id and topic

JenySadadia commented:
@mgalka  Can you please confirm that I have used the correct values for KCIDB configurations:
https://github.com/kernelci/kernelci-pipeline/pull/188/files#diff-f1c9e56ef349bb697df7f6aab238aee7e28615282824ad4c8862335adbaa69b7

gctucker commented on src/send_kcidb.py:
```diff
@@ -8,126 +8,122 @@
+        self.log.info("Press Ctrl-C to stop.")
         sys.stdout.flush()
 
+        client = self._kcidb_connect()
+        if not client:
+            self.log.error("Failed to create client connection to KCIDB")
+            return False
```
I think it would be better to do this in `self._setup()`.  Then any disconnection can be handled when an exception happens during the main loop, or the loop could be restarted without redoing the KCIDB connection etc.

JenySadadia replied:
Yes, makes sense.

gctucker commented on config/kernelci.conf:
```diff
@@ -20,7 +20,9 @@ output: /home/kernelci/output
-[kcidb]
+[send_kcidb]
+kcidb_topic_name: playground_kcidb_new
+kcidb_project_id: kernelci-production
```
Not sure this is the right project id, first of all the pipeline is actually running on staging.  Maybe `kernelci-api` would be better at this stage to differentiate it from the data sent by the current actual production / staging setup which uses `kernelci_backend`.

JenySadadia replied:
Updated the project ID.

gctucker commented on src/send_kcidb.py:
```diff
@@ -8,126 +8,116 @@
-                }
-                self.send_revision(client, revision)
+    def _stop(self, sub_id):
+        if sub_id:
```
I guess the KCIDB Client object doesn't have any means of disconnecting?

If it's held in the context object then it will go out of scope when the loop exits and any clean-up will be done automatically then.

JenySadadia replied:
I checked `kcidb.Client` implementation. It doesn't have any external method to disconnect.

gctucker commented on src/send_kcidb.py:
```diff
@@ -8,126 +8,116 @@
+        if self._validate_revision(revision):
+            return client.submit(revision)
+        self.log.error("Aborting, invalid data")
+        sys.stdout.flush()
```
I think we can drop all the `.flush()` calls now.

gctucker commented on src/send_kcidb.py:
```diff
@@ -8,126 +8,116 @@
-        super().__init__(sub_parser, name)
-        self._logger = Logger("config/logger.conf", "send_kcidb")
+    def _kcidb_connect(self):
+        return Client(project_id=self._kcidb_project_id,
+                      topic_name=self._kcidb_topic_name)
```
I mean, the `Client` object could be created directly in `_setup()`.  Also the KCIDB topic and project ID could be read from the `args` there rather than storing them as instance variables in the object.  That's a bit more of a grey area but I think it's better to set everything up in `_setup()` as much as possible.

JenySadadia replied:
Having separate functions increases the code's reusability.
That's why I created the `kcidb_connect` method and also verified if the connection is successful in the method.
Since the method just has a single line of code now, it won't make much of a difference and the method can be dropped.

gctucker commented on src/send_kcidb.py:
```diff
@@ -8,126 +8,113 @@
-            self._logger.log_message(logging.ERROR, traceback.format_exc())
-        finally:
-            db.unsubscribe(sub_id)
+    @staticmethod
+    def _validate_revision(revision):
+        return kcidb.io.SCHEMA.is_valid(revision)
```
I think this method can be dropped, we can just call `.is_valid()` directly.

gctucker commented:
OK let's just wait for some staging results now to check it's working as expected.

mgalka commented:
> @mgalka Can you please confirm that I have used the correct values for KCIDB configurations? https://github.com/kernelci/kernelci-pipeline/pull/188/files#diff-f1c9e56ef349bb697df7f6aab238aee7e28615282824ad4c8862335adbaa69b7

These look good

gctucker commented:
There's a problem though:
```
google.api_core.exceptions.NotFound: 404 Requested project not found or user does not have access to it (project=kernelci-api). Make sure to specify the unique project identifier and not the Google Cloud Console display name.
```

JenySadadia commented:
> There's a problem though:
> 
> ```
> google.api_core.exceptions.NotFound: 404 Requested project not found or user does not have access to it (project=kernelci-api). Make sure to specify the unique project identifier and not the Google Cloud Console display name.
> ```

I have modified project id and added `misc` field to identify submitter.

JenySadadia commented:
As per our discussion with @spbnick, I have changed `origin` to `kernelci-api` to segregate staging results sent from the new API.

gctucker commented:
I've merged the kernelci-deploy PR so let's wait for the next staging run to confirm it's sending the data as expected.

JenySadadia commented:
Seems like an issue with submission:
```
today at 06:48:1812/02/2022 01:18:18 AM UTC [INFO] Submitting node to KCIDB: 638944495511e7db63a0f99f
today at 06:48:1812/02/2022 01:18:18 AM UTC [ERROR] Aborting, invalid data
```
Added a debug print to verify the revision data.

gctucker commented:
OK great, at least it's running and talking to KCIDB.  I guess the data being sent doesn't match the required schema.

gctucker commented on src/send_kcidb.py:
```diff
@@ -8,126 +8,109 @@
+                'builds': [],
+                'checkouts': [{
+                    'id': f"kernelci:{node['_id']}",
+                    'origin': 'kernelci-api',
```
Note: I think it would be good to have a `--origin` command line option, as a follow-up.

JenySadadia replied:
Let me do it in this PR itself as `origin` has been added in this one.

gctucker replied:
The `origin` field was already there as `kernelci`, but yes this PR is changing it.  I guess we can leave `kernelci` by default but set `kernelci-api` in the staging configuration file in `kernelci-deploy`.

gctucker replied:
Ah I see you've already added it.  It might have been better as a separate commit as that's more than just refactoring the code to use the Service base class.

JenySadadia replied:
Yes, I'll add a separate commit.

JenySadadia replied:
Also, corrected `checkouts.id` field to use `origin` in the value to fix invalid data error.

JenySadadia commented:
Working on staging:
```
today at 06:34:4212/05/2022 01:04:42 AM UTC [INFO] Submitting node to KCIDB: 638d359a5511e7db63a0f9ae
today at 06:34:4212/05/2022 01:04:42 AM UTC [DEBUG] Submitting revision: {'builds': [], 'checkouts': [{'id': 'kernelci:638d359a5511e7db63a0f9ae', 'origin': 'kernelci', 'tree_name': 'kernelci', 'git_repository_url': 'https://github.com/kernelci/linux.git', 'git_commit_hash': '4009eec93bfa88b8da68864e8c3c65808d437eec', 'git_repository_branch': 'staging-mainline', 'start_time': '2022-12-05T00:04:42.061000+00:00', 'patchset_hash': '', 'misc': {'submitted_by': 'kernelci-pipeline'}}], 'tests': [], 'version': {'major': 4, 'minor': 0}}
```

The only issue is the value of `origin`. Its value from settings file will be overridden by the `default` value of the command line option.
Fixed it for `origin` to be `kernelci-api`.

gctucker commented:
> The only issue is the value of `origin`. Its value from settings file will be overridden by the `default` value of the command line option.
> Fixed it for `origin` to be `kernelci-api`.

If we can't set the value in the settings file due to the default value in the command line argument definition, then we just need to remove the default value from the command line argument definition.  I guess it makes sense that it takes precedence as the code treats the default command line argument value as a value passed by the user, which is meant to override the values in the settings file.

We can deal with a default value in the application code outside of the argument definition, to still set it to `kernelci` when `args.origin` is empty.

JenySadadia commented:
> > The only issue is the value of `origin`. Its value from settings file will be overridden by the `default` value of the command line option.
> > Fixed it for `origin` to be `kernelci-api`.
> 
> If we can't set the value in the settings file due to the default value in the command line argument definition, then we just need to remove the default value from the command line argument definition. I guess it makes sense that it takes precedence as the code treats the default command line argument value as a value passed by the user, which is meant to override the values in the settings file.
> 
> We can deal with a default value in the application code outside of the argument definition, to still set it to `kernelci` when `args.origin` is empty.

Or else we can keep the default value in the command line argument definition for staging and we can drop it later. We anyway have to use `origin:kernelci` when it gets deployed to production.

gctucker commented:
How about setting `kernelci_api` in the `staging.kernelci.org.conf` file and no default value?  That way we won't have to change the code for production or if anyone else wants to set up a pipeline and send data to KCIDB.

JenySadadia commented:
> How about setting `kernelci_api` in the `staging.kernelci.org.conf` file and no default value? That way we won't have to change the code for production or if anyone else wants to set up a pipeline and send data to KCIDB.

Seems a better idea. I will keep the origin `kernelci` in settings file for main branch and `kernelci_api` for staging.

JenySadadia commented:
Created https://github.com/kernelci/kernelci-deploy/pull/41

JenySadadia commented:
Tested OK on staging with `origin:kernelci_api`:

KCIDB bridge:
```
today at 06:44:1312/07/2022 01:14:13 AM UTC [INFO] Submitting node to KCIDB: 638fdad44f2e91321f7c31ff
today at 06:44:1312/07/2022 01:14:13 AM UTC [DEBUG] Submitting revision: {'builds': [], 'checkouts': [{'id': 'kernelci_api:638fdad44f2e91321f7c31ff', 'origin': 'kernelci_api', 'tree_name': 'kernelci', 'git_repository_url': 'https://github.com/kernelci/linux.git', 'git_commit_hash': 'e3ab23ecb5c8f2b0bcd91e48f152d7b3f9ac6744', 'git_repository_branch': 'staging-stable', 'start_time': '2022-12-07T00:14:12.429000+00:00', 'patchset_hash': '', 'misc': {'submitted_by': 'kernelci-pipeline'}}], 'tests': [], 'version': {'major': 4, 'minor': 0}}
```

Link to [KCIDB dashboard](https://kcidb.kernelci.org/d/checkout/checkout?orgId=1&var-dataset=playground_kcidb_01&var-origin=kernelci_api&var-build_architecture=All&var-build_config_name=All&var-id=kernelci_api:638fdad44f2e91321f7c31ff)
